server/live-loan-router.js

```diff
@@ -3,170 +3,203 @@ const express = require('express');
+					country {
+						name
+					}
+				}
+				use
+				loanAmount
+				status
+				loanFundraisingInfo {
+					fundedAmount
+				}
+				image {
+					retina: url(customSize: "w960h720")
+				}
 			}
-			resolve(data);
-		});
-	});
-}
+		}
+	}
+}`;
 
-function setToCache(key, value, expires, cache) {
-	return new Promise((resolve, reject) => {
-		cache.set(key, value, { expires }, (error, success) => {
-			if (error) {
-				console.error(JSON.stringify({
-					meta: {},
-					level: 'error',
-					message: `MemJS Error Setting Cache for ${key}, Error: ${error}`
-				}));
-				reject();
-			}
-			if (success) {
-				console.info(JSON.stringify({
-					meta: {},
-					level: 'info',
-					message: `MemJS Success Setting Cache for ${key}, Success: ${success}`
-				}));
-				resolve();
+const recommendationsByLoanIdQuery = id => `{
+	ml {
+		relatedLoansByTopics(
+			loanId: ${id},
+			offset: 0,
+			topics: [story]
+			limit: 4,
+		) {
+			values {
+				name
+				id
+				borrowerCount
+				geocode {
+					country {
+						name
+					}
+				}
+				use
+				loanAmount
+				status
+				loanFundraisingInfo {
+					fundedAmount
+				}
+				image {
+					retina: url(customSize: "w960h720")
+				}
 			}
+		}
+	}
+}`;
+
+async function fetchRecommendedLoans(type, id, cache) {
+	// If we have loan data in memjscache return that quickly
+	try {
+		const cachedLoanData = await memJsUtils.getFromCache(`recommendations-by-${type}-id-${id}`, cache);
+		if (cachedLoanData) {
+			return JSON.parse(cachedLoanData);
+		}
+	} catch (err) {
+		log(`Error reading from memjs, ${err}`, 'error');
+	}
+
+	// Otherwise we need to hit the graphql endpoint.
+	// prep the query
+	let query;
+	let queryResultPath;
+	if (type === 'user') {
+		query = recommendationsByLoginIdQuery(id);
+		queryResultPath = 'data.ml.recommendationsByLoginId.values';
+	} else if (type === 'loan') {
+		query = recommendationsByLoanIdQuery(id);
+		queryResultPath = 'data.ml.relatedLoansByTopics[0].values';
+	} else {
+		throw new Error('Type must be user or loan');
+	}
+
+	// hit the endpoint and parse the response
+	let loanData = [];
+	try {
+		const endpoint = config.app.graphqlUri;
+		const response = await fetch(endpoint, {
+			method: 'POST',
+			headers: { 'Content-Type': 'application/json' },
+			body: JSON.stringify({
+				query
+			}),
 		});
-	});
+		const result = await response.json();
+		loanData = get(result, queryResultPath);
+	} catch (err) {
+		log(`Error fetching loans: ${err}`, 'error');
+	}
+
+	// Set the loan data in memcache, return the loan data
+	if (loanData.length) {
+		try {
+			const expires = 10 * 60; // 10 minutes
+			await memJsUtils.setToCache(
+				`recommendations-by-${type}-id-${id}`,
+				JSON.stringify(loanData),
+				expires,
+				cache
+			);
+			return loanData;
+		} catch (err) {
+			log(`Error setting loan data to cache, ${err}`, 'error');
+		}
+	} else {
+		// future improvement, return a default set of 4 loans here
+		throw new Error('No loans returned');
```

Comment:
@mcstover I'm throwing here to reject out of the promise, which gets caught and logged in `redirectToUrl` and `serveImg`. An alternative is to resolve as a blank array, and then check that we have loans before processing on lines 143 and 159.